File: DFS_BFS/468-populateRightOfBinaryTree.py
"""
given a binary tree:
struct Node {
  int val;
  Node *left;
  Node *right;
  Node *next;
}
Populate each next pointer to point to its next right node. 
If there is no next right node, the next pointer should be set to NULL. Initially, 
all next pointers are set to NULL.
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque
class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if not root:
            return root
        q=deque([root])
        while q:
            n=len(q)
            for i in range(n):
                cur=q.popleft()
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
                if i!=n-1:
                    cur.next=q[0]
        return root

File: DFS_BFS/468-populateRightOfBinaryTree.cpp
#include<iostream>
#include<deque>

class TreeNode
{
    private:
        int m_val{};
        TreeNode* left{};
        TreeNode* right{};
        TreeNode* next{};
    public:
        TreeNode(int v, TreeNode* left=nullptr, TreeNode* right=nullptr,TreeNode* next=nullptr):
            m_val{v},left{left},right{right},next{next}{};
        TreeNode* getRight() const {return right;}
        void setNext(TreeNode* node){next=node;}
};

using Q=std::queue<TreeNode>;

TreeNode* populateRight(TreeNode& root)
{
    if(!root)
        return root;
    Q q{root};
    while(!q.empty())
    {
        int n=q.size();
        for(int i=0;i<n;++i)
        {
            TreeNode* cur=q.front();
            q.pop_front();
            if(cur->left)
                q.push_back(cur->left);
            if(cur->right)
                q.push_back(cur->right);
            if(i!=n-1)
                cur->setNext(q.front());
        }
    }
    return root;
}
